torch/utils/_pytree.py

```diff
@@ -470,14 +470,23 @@ def _is_leaf(tree: PyTree) -> bool:
 # context: some context that is useful in unflattening the pytree
 # children_specs: specs for each child of the root Node
 # num_leaves: the number of leaves
-@dataclasses.dataclass
+@dataclasses.dataclass(init=True, frozen=True, eq=True, repr=False)
 class TreeSpec:
     type: Any
     context: Context
     children_specs: List["TreeSpec"]
 
+    num_nodes: int = dataclasses.field(init=False)
+    num_leaves: int = dataclasses.field(init=False)
+    num_children: int = dataclasses.field(init=False)
+
     def __post_init__(self) -> None:
-        self.num_leaves: int = sum([spec.num_leaves for spec in self.children_specs])
+        num_nodes = 1 + sum(spec.num_nodes for spec in self.children_specs)
+        num_leaves = sum(spec.num_leaves for spec in self.children_specs)
+        num_children = len(self.children_specs)
+        object.__setattr__(self, "num_nodes", num_nodes)
+        object.__setattr__(self, "num_leaves", num_leaves)
+        object.__setattr__(self, "num_children", num_children)
```

Comment:
Dynamo does not support `object.__setattr__(obj, name, value)` yet.

For all frozen dataclasses, the assignment statement in `__init__` are `object.__setattr__(obj, name, value)` rather than `obj.<name> = value`.

https://github.com/python/cpython/blob/05a370abd6cdfe4b54be60b3b911f3a441026bb2/Lib/dataclasses.py#L477-L486

```python
def _field_assign(frozen, name, value, self_name):
    # If we're a frozen class, then assign to our fields in __init__
    # via object.__setattr__.  Otherwise, just use a simple
    # assignment.
    #
    # self_name is what "self" is called in this function: don't
    # hard-code "self", since that might be a field name.
    if frozen:
        return f'__dataclass_builtins_object__.__setattr__({self_name},{name!r},{value})'
    return f'{self_name}.{name}={value}'
```

The error:

```text
torch._dynamo.exc.Unsupported: call_method BuiltinVariable(object) __setattr__ [UserDefinedObjectVariable(TreeSpec), ConstantVariable(str), ConstantVariable(int)] {}
```

A minimal script to reproduce:

```python
import dataclasses

import torch


@dataclasses.dataclass
class A:
    a: int


def fn() -> None:
    a = A(1)
    object.__setattr__(a, "a", 2)


comp_out = torch._dynamo.optimize(nopython=True)(fn)()
```